### PR TITLE
Make UnrollCustomDefinitions pass target aware

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

This commit makes the UnrollCustomDefinitions transpiler pass target aware. A new kwarg, target is added to the pass constructor and if specified the Target object passed in is used to check if the target backend supports the gates in the circuit instead of the basis_gates list.

### Details and comments

Part of #7113

### CODE REVIEW

File: qiskit/transpiler/passes/basis/unroll_custom_definitions.py [hunk 1]
```diff
@@ -67,11 +73,22 @@ def run(self, dag):
...
+                        pass
+                    else:
+                        continue
+
+            else:
+                if node.name in device_insts or self._equiv_lib.has_entry(node.op):
+                    if isinstance(node.op, ControlledGate) and node.op._open_ctrl:
+                        pass
+                    else:
+                        continue
```
Comment (kevinhartman) [inline qiskit/transpiler/passes/basis/unroll_custom_definitions.py:91]:
```suggestion
            controlled_gate_open_ctrl = isinstance(node.op, ControlledGate) and node.op._open_ctrl
            if not controlled_gate_open_ctrl:
                inst_supported = self._target.instruction_supported(
                    operation_name=node.op.name, qargs=tuple(qubit_mapping[x] for x in node.qargs)
                ) if self._target is not None else node.name in device_insts

                if inst_supported or self._equiv_lib.has_entry(node.op):
                    continue
```

Untested (not sure if Python will be happy with `inst_supported` statement...), but hopefully something along these lines will improve readability.

File: releasenotes/notes/target-aware-unroll-custom-definitions-a1b839de199ca048.yaml [hunk 2]
```diff
@@ -0,0 +1,8 @@
+---
+features:
+  - |
+    The :class:`~.UnrollCustomDefinitions` transpiler pass has a new keyword
+    argument, ``target``, on it's constructor. This argument can be used to
```
Comment (kevinhartman) [inline releasenotes/notes/target-aware-unroll-custom-definitions-a1b839de199ca048.yaml:5]:
```suggestion
    argument, ``target``, on its constructor. This argument can be used to
```

**Review by kevinhartman — COMMENTED** [review]
LGTM. Added a few cleanup suggestions.

File: qiskit/transpiler/passes/basis/unroll_custom_definitions.py [hunk 3]
```diff
@@ -50,11 +54,13 @@ def run(self, dag):
             DAGCircuit: output unrolled dag
         """
 
-        if self._basis_gates is None:
+        if self._basis_gates is None and self._target is None:
             return dag
```
[the remark this answers is not in the document]
Comment (kevinhartman) [reply qiskit/transpiler/passes/basis/unroll_custom_definitions.py:58]: Ah, that's right. I forgot about this requirement.

File: qiskit/transpiler/passes/basis/unroll_custom_definitions.py (hunk 1 above)
Comment (mtreinish) [reply qiskit/transpiler/passes/basis/unroll_custom_definitions.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/8784/commits/d87ce39027844b16a3a449a6ed97ca6fdb2f87c1

**Review by kevinhartman — APPROVED** [review]
LGTM!

**Pull request merged**